scalableminds/webknossos
# Fix race condition in finishAndGetNextTask

When using the finishAndGetNextTask method from our API, we make sure to save the tracing (which empties the save queue), before finishing the task and then hot-swapping to the next task. However if the user creates a node in the time between the saving and hot-swapping (while the finishTask request is executed), this node is pushed to the save queue and will be saved as part of the new hot-swapped task.
This is of course wrong and led to invalid tracings.

My solution is to empty the save queue after the task hot swap was completed.

There was another minor issue with the `model.save` method. This method is usually used like this: `await Model.save()`. It triggers the `saveNowAction` and then repeatedly (every 500ms) checks whether the save queue is now empty. If however a new change was pushed to the save queue in those 500ms, the while loop will run indefinitely (until the next periodic save is triggered). Therefore I moved the dispatch of the `saveNowAction` into the while loop. Calling the `saveNowAction` repeatedly doesn't do any harm and now it is made sure that `await Model.save()` will not need to wait for 30s in some cases.

### Mailable description of changes:
 - Fixed a race condition in `finishAndGetNextTask`, that led to invalid tracings.

### Steps to test:
- The easiest way to trigger the first bug is to create tasks that load this script: https://gist.github.com/kevinbbb87/ac0dfdf1a43fc27aa3b5a86561c2108b
- Set some nodes, then use `9` to trigger the task hot swap, wait until the checkmark on the save button appears and then immediately set another node (while the finish task request runs). Then in the hot-swapped task, set a node and press save. The node from the old tracing, should not be part of the update actions that are sent to the server.

- The second bug could be triggered by tracing in flight mode and then pressing `9` to swap the task (while still pressing `space`), then release space. The task should be swapped immediately after releasing `space` (before nothing happened after releasing `space`, because `finishAndGetNextTask` was still stuck in `await Model.save()`).

### Issues:
- fixes #2075 

------
